extensions/amp-base-carousel/1.0/base-carousel.js

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as Preact from '../../../src/preact';
+import {Arrow} from './arrow';
+import {Scroller} from './scroller';
+import {toChildArray, useRef, useState} from '../../../src/preact';
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function BaseCarousel(props) {
+  const {
+    'arrowPrev': arrowPrev,
+    'arrowNext': arrowNext,
+    'children': children,
+    'currentSlide': currentSlide,
+    'loop': loop,
+    ...rest
+  } = props;
+  const {length} = toChildArray(children);
+  const {0: curSlide, 1: setCurSlide} = useState(currentSlide || 0);
+  const ignoreProgrammaticScroll = useRef(true);
+  const setRestingIndex = (i) => {
+    ignoreProgrammaticScroll.current = true;
+    setCurSlide(i);
+  };
+  const scrollRef = useRef(null);
+  const advance = (dir) => {
```

Comment:
At a high level I would agree with you -- this is, in fact, a programmatic scroll. However it's not present here simply because it is not effective even if we do disable it.

To get the smooth scroll, I set `scrollLeft` and rely on the `scrollBehavior: smooth` styling. The disabling makes the scroll handler a noop for exactly the next scroll event. Because a programmatic smooth scroll triggers many scroll events throughout the duration of the scroll, this would not effectively disable the scroll handler for the full length of scroll. 

The reason I don't disable programmatic scroll and `setState` right from the start -- the render triggered by doing this negates the smooth scroll visual. So instead it triggers a non-disabled programmatic scroll that implicitly relies on the debounced scroll handler to catch and update the state as a consequence of the scroll. I agree it's somewhat funny. If you have any feedback on how to make this logic feel more intuitive I'd appreciate it :)